predicators/envs/sandwich.py

```diff
@@ -0,0 +1,832 @@
+            y = state.get(obj, "pose_y") - length / 2.
+            z = state.get(obj, "pose_z") - thickness / 2.
+            return utils.Rectangle(y, z, length, thickness, 0.0)
+        # Cylinder.
+        assert abs(state.get(obj, "shape") - 1.0) < 1e-3
+        thickness = state.get(obj, "thickness")
+        radius = state.get(obj, "radius")
+        length = 2 * radius
+        # Oriented facing up, i.e., in the holder.
+        if abs(state.get(obj, "rot") - np.pi / 2.) < 1e-3:
+            y = state.get(obj, "pose_y") - thickness / 2.
+            z = state.get(obj, "pose_z") - length / 2.
+            return utils.Rectangle(y, z, thickness, length, 0.0)
+        # Oriented facing down, i.e., on the board.
+        assert abs(state.get(obj, "rot") - 0.0) < 1e-3
+        y = state.get(obj, "pose_y") - length / 2.
+        z = state.get(obj, "pose_z") - thickness / 2.
+        return utils.Rectangle(y, z, length, thickness, 0.0)
+
+    def _obj_to_color(self, obj: Object, state: State) -> RGBA:
+        if obj.is_instance(self._holder_type):
+            return self.holder_color
+        if obj.is_instance(self._board_type):
+            return self.board_color
+        assert obj.is_instance(self._ingredient_type)
+        r = state.get(obj, "color_r")
+        g = state.get(obj, "color_g")
+        b = state.get(obj, "color_b")
+        a = 1.0
+        return (r, g, b, a)
+
+    def _ingredient_to_static_features(self,
+                                       ing_name: str) -> Dict[str, float]:
+        color_r, color_g, color_b = self.ingredient_colors[ing_name]
+        radius = self.ingredient_radii[ing_name]
+        shape = self.ingredient_shapes[ing_name]
+        return {
+            "color_r": color_r,
+            "color_g": color_g,
+            "color_b": color_b,
+            "thickness": self.ingredient_thickness,
+            "radius": radius,
+            "shape": shape
+        }
+
+    def _is_ingredient(self, obj: Object, ing_name: str, state: State) -> bool:
+        return self._obj_to_ingredient(obj, state) == ing_name
+
+    def _obj_to_ingredient(self, obj: Object, state: State) -> str:
+        obj_color = (state.get(obj, "color_r"), state.get(obj, "color_g"),
+                     state.get(obj, "color_b"))
+        affinities = {
+            n: np.sum(np.subtract(c, obj_color)**2)
+            for n, c in self.ingredient_colors.items()
+        }
+        closest = min(affinities, key=lambda o: affinities[o])
+        return closest
+
+    def _state_to_ingredient_groups(self,
+                                    state: State) -> Dict[str, Set[Object]]:
+        ings = set(self.ingredient_colors)
+        ing_groups: Dict[str, Set[Object]] = {n: set() for n in ings}
+        for obj in state.get_objects(self._ingredient_type):
+            ing_name = self._obj_to_ingredient(obj, state)
+            ing_groups[ing_name].add(obj)
+        return dict(ing_groups)
```

Comment:
okay, this helps me see a bit why the `_obj_to_ingredient` function is written as above, but I'm still quite confused why you're using the color...